### PR TITLE
fix(raft): reset snapshot index on failure

### PR DESCRIPTION
**Pull request by npepinpe**
- resets the snapshotIndex of a member to 0 (initial state) when we reset its index

### CODE REVIEW

**Comment by npepinpe** [comment]
Not exactly sure how I'm going to test this in the Atomix project...

**Comment by npepinpe** [comment]
@deepthidevaki I will try to add more tests at a later date, but if you think this looks fine then I would merge this (unless you're not comfortable with that, which is totally valid since we'll release this while I'm on holiday :sweat_smile: )

File: protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java
Comment (deepthidevaki) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java]: I am wondering if it makes more sense to encode `lastSnapshotIndex` of `member` in the response. `resetSnapshotIndex(member, response.lastSnapshotIndex())`. In case of full data loss this value will be 0. Otherwise it will be the actual snapshot index.

**Comment by deepthidevaki** [comment]
I have an idea how to test this. We can adapt `io.atomix.protocols.raft.protocol.TestRaftServerProtocol` to record each install or append requests received by a node. Add a test to `io.atomix.protocols.raft.protocol.RaftTest` where a node is restarted after deleting its data directory and inspects its TestRaftServerProtocol instance for received install requests. (I tried it  [here](https://github.com/zeebe-io/atomix/blob/775bcd06fed698fc296ce2b3f196d309d9477919/protocols/raft/src/test/java/io/atomix/protocols/raft/protocol/TestRaftServerProtocol.java#L427)). It might be a bit of work though.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java
Comment (deepthidevaki) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java]: So I did a small test to see how resetting `snapshotIndex` to 0 affect the scenario where there is no data loss. The setup has a slow server, which might lose some append requests. In this case there is no restart. Just that one follower(network connection) is slow. In this case, I observed that the leader sends the same snapshot multiple times to this slow follower. It doesn't lead to any inconsistent state. The follower is capable of handling redundant requests. But it can unnecessarily put additional load on the already slow network to this follower. So I would like to see whether the above approach I mentioned works.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]

File: protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java
Comment (npepinpe) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java]: I thought about that, but I figured this would be a more acceptable change for the maintainers, as it's pretty unobstrusive; I figured we're adding a field to the response which is only used some times, so they might not accept it. I admit that's kind of a weak argument though :sweat_smile:

**Comment by npepinpe** [comment]
I updated it to actually reset to the remote member's last snapshot index.

We discussed it over slack and I think we would have a good case to make via upstream as to why it makes sense to modify the `AppendResponse` to include the member's last snapshot index.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/protocol/AppendResponse.java
Comment (deepthidevaki) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/protocol/AppendResponse.java]:
```suggestion
    return Objects.hash(getClass(), status, term, succeeded, lastLogIndex, lastSnapshotIndex);
```

File: protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java
Comment (deepthidevaki) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/roles/AbstractAppender.java]:
```suggestion
   * Resets the snapshot index of the member when a response fails.
```

**Review by deepthidevaki — CHANGES_REQUESTED** [review]

**Review by deepthidevaki — APPROVED** [review]

**Pull request merged**
